pelicun/model/loss_model.py

```diff
@@ -1210,72 +1241,101 @@ def _aggregate_sample(self, sample: pd.DataFrame, columns: list) -> pd.DataFrame
+            tuple[uq.RandomVariableRegistry, dict[str, float]] | None
+        ),
     ) -> tuple[pd.DataFrame, pd.DataFrame]:
 
-        # If there are no `replacement_thresholds`, simply return.
-        if replacement_thresholds is None or not replacement_thresholds:
+        # If there is no `replacement_configuration`, simply return.
+        if replacement_configuration is None:
             # `exceedance_bool_df` is empty in this case.
             exceedance_bool_df = pd.DataFrame(index=sample.index, dtype=bool)
             return sample, exceedance_bool_df
 
-        # otherwise we initialize it with False
+        replacement_consequence_RV_reg, replacement_ratios = (
+            replacement_configuration
+        )
         exceedance_bool_df = pd.DataFrame(
             False,
             index=sample.index,
-            columns=replacement_thresholds.keys(),
+            columns=replacement_consequence_RV_reg.RV.keys(),
             dtype=bool,
         )
 
-        # If there is no `replacement` in the sample, simply return.
-        if 'replacement' not in sample.columns.get_level_values('loss'):
-            return sample, exceedance_bool_df
+        # Sample replacement consequences from the registry
+        replacement_consequence_RV_reg.generate_sample(len(sample), 'MonteCarlo')
+
+        sample_dvs = replacement_consequence_RV_reg.RV.keys()
+        for sample_dv in sample_dvs:
+            sub_sample = sample.loc[:, sample_dv]
+            if 'replacement' in sub_sample.columns.get_level_values('loss'):
+                # If `replacement` already exists as a consequence,
+                # determine the realizations where it is non-zero.
+                no_replacement_mask = (
+                    ~(sub_sample['replacement'] > 0.00).any(axis=1).values
+                )
+                no_replacement_columns = (
+                    sub_sample.columns.get_level_values('loss') != 'replacement'
+                )
+            else:
+                # Otherwise there is no row where we already have replacement
+                no_replacement_mask = np.full(len(sub_sample), True)
+                no_replacement_columns = np.full(len(sub_sample.columns), True)
+            # Get the sum to compare with the thresholds
+            consequence_sum_given_no_replacement = sub_sample.iloc[
+                no_replacement_mask, no_replacement_columns
+            ].sum(axis=1)
+            if not consequence_sum_given_no_replacement.empty:
+                consequence_values = replacement_consequence_RV_reg.RV[
+                    sample_dv
+                ].sample
+                exceedance_mask = (
+                    consequence_sum_given_no_replacement
+                    > consequence_values[no_replacement_mask]
+                    * replacement_ratios[sample_dv]
+                )
+                # Fill the remaining rows with False
+                if len(exceedance_mask) < len(sub_sample):
```

Comment:
Why would they not be the same length? And why is it a good idea to reindex and Fill in that case?

Reply:
`no_replacement_mask` is a boolean index, so `consequence_sum_given_no_replacement` can end up with fewer rows in this line, or even empty:
```{python}
            consequence_sum_given_no_replacement = sub_sample.iloc[
                no_replacement_mask, no_replacement_columns
            ].sum(axis=1)
```
The missing rows correspond to realizations where there is no replacement.
```
                    exceedance_mask = exceedance_mask.reindex(
                        sub_sample.index, fill_value=False
                    )
```
The above makes the index of `exceedance_mask` match that of `sub_sample`, and assigns `False` to the missing rows.

Reply:
Makes sense, thanks.